src/core/qgsogrutils.cpp

```diff
@@ -2034,6 +2035,124 @@ QVariant QgsOgrUtils::stringToVariant( OGRFieldType type, OGRFieldSubType, const
   return ok ? res : QVariant();
 }
 
+QList<QgsVectorDataProvider::NativeType> QgsOgrUtils::nativeFieldTypesForDriver( GDALDriverH driver )
+{
+  if ( !driver )
+    return {};
+
+  const QString driverName = QString::fromUtf8( GDALGetDriverShortName( driver ) );
+
+  int nMaxIntLen = 11;
+  int nMaxInt64Len = 21;
+  int nMaxDoubleLen = 20;
+  int nMaxDoublePrec = 15;
+  int nDateLen = 8;
+  if ( driverName == QLatin1String( "GPKG" ) )
+  {
+    // GPKG only supports field length for text (and binary)
+    nMaxIntLen = 0;
+    nMaxInt64Len = 0;
+    nMaxDoubleLen = 0;
+    nMaxDoublePrec = 0;
+    nDateLen = 0;
+  }
+
+  QList<QgsVectorDataProvider::NativeType> nativeTypes;
+  nativeTypes
+      << QgsVectorDataProvider::NativeType( QgsVariantUtils::typeToDisplayString( QVariant::Int ), QStringLiteral( "integer" ), QVariant::Int, 0, nMaxIntLen )
+      << QgsVectorDataProvider::NativeType( QgsVariantUtils::typeToDisplayString( QVariant::LongLong ), QStringLiteral( "integer64" ), QVariant::LongLong, 0, nMaxInt64Len )
+      << QgsVectorDataProvider::NativeType( QObject::tr( "Decimal number (real)" ), QStringLiteral( "double" ), QVariant::Double, 0, nMaxDoubleLen, 0, nMaxDoublePrec )
+      << QgsVectorDataProvider::NativeType( QgsVariantUtils::typeToDisplayString( QVariant::String ), QStringLiteral( "string" ), QVariant::String, 0, 65535 );
+
+  if ( driverName == QLatin1String( "GPKG" ) )
+    nativeTypes << QgsVectorDataProvider::NativeType( QObject::tr( "JSON (string)" ), QStringLiteral( "JSON" ), QVariant::Map, 0, 0, 0, 0, QVariant::String );
+
+  bool supportsDate = true;
```

Comment:
This remark doesn't specifically apply to this PR as you just moved around existing code, but supportsDate, supportsTime and supportsDateTime should be likely all initialized to true, because the GPKG and Shapefile drivers have GDAL_DMD_CREATIONFIELDDATATYPES so they don't need specific initializations.
Initializing to true will keep existing behaviour unchanged, although it is a bit arguable that's the appropriate default

Reply:
Thanks, done